From c48fe30145b274b4d5ccee7d9f67b48650441348 Mon Sep 17 00:00:00 2001
From: Vladimir Barinov <vladimir.barinov@cogentembedded.com>
Date: Mon, 27 May 2013 10:07:21 +0900
Subject: [PATCH 8/9] Add VIN clocks and platform devices on R8A7778 SoC; add
 function to register the VIN platform devices.

Signed-off-by: Vladimir Barinov <vladimir.barinov@cogentembedded.com>
[Sergei: added 'id' parameter check to r8a7779_add_vin_device(), used '*pdata'
in *sizeof* operator there, renamed some variables, annotated 'vin[01]_info' and
vin[01]_resources[] as '__initdata'.]
Signed-off-by: Sergei Shtylyov <sergei.shtylyov@cogentembedded.com>
---
 arch/arm/mach-shmobile/clock-r8a7778.c        |  5 ++++
 arch/arm/mach-shmobile/include/mach/r8a7778.h |  4 ++++
 arch/arm/mach-shmobile/setup-r8a7778.c        | 34 +++++++++++++++++++++++++++
 3 files changed, 43 insertions(+)

diff --git a/arch/arm/mach-shmobile/clock-r8a7778.c b/arch/arm/mach-shmobile/clock-r8a7778.c
index 3856e60..26fc558 100644
--- a/arch/arm/mach-shmobile/clock-r8a7778.c
+++ b/arch/arm/mach-shmobile/clock-r8a7778.c
@@ -107,6 +107,7 @@ static struct clk div4_clks[DIV4_NR] = {
 enum { MSTP517, MSTP513, MSTP501,
 	MSTP410, MSTP408,
 	MSTP331, MSTP323, MSTP319,
+	MSTP110, MSTP109,
 	MSTP107, MSTP105, MSTP103, MSTP100,
 	MSTP030,
 	MSTP029, MSTP028, MSTP027, MSTP026, MSTP025, MSTP024, MSTP023, MSTP022, MSTP021,
@@ -123,6 +124,8 @@ static struct clk mstp_clks[MSTP_NR] = {
 	[MSTP331] = SH_CLK_MSTP32(&div4_clks[DIV4_S4], MSTPCR3, 31, 0), /* MMC */
 	[MSTP323] = SH_CLK_MSTP32(&div4_clks[DIV4_P], MSTPCR3, 23, 0), /* SDHI0 */
 	[MSTP319] = SH_CLK_MSTP32(&div4_clks[DIV4_P], MSTPCR3, 19, 0), /* SCR */
+	[MSTP110] = SH_CLK_MSTP32(&div4_clks[DIV4_S], MSTPCR1, 10, 0), /* VIN0 */
+	[MSTP109] = SH_CLK_MSTP32(&div4_clks[DIV4_S], MSTPCR1,  9, 0), /* VIN1 */
 	[MSTP107] = SH_CLK_MSTP32(&div4_clks[DIV4_P], MSTPCR1,  7, 0), /* GP2D */
 	[MSTP105] = SH_CLK_MSTP32(&div4_clks[DIV4_S], MSTPCR1,  5, 0), /* SGX */
 	[MSTP103] = SH_CLK_MSTP32(&div4_clks[DIV4_S], MSTPCR1,  3, 0), /* DU */
@@ -165,6 +168,8 @@ static struct clk_lookup lookups[] = {
 	CLKDEV_DEV_ID("sh_mmcif",	&mstp_clks[MSTP331]), /* MMC */
 	CLKDEV_DEV_ID("sh_mobile_sdhi.0",	&mstp_clks[MSTP323]), /* SDHI0 */
 	CLKDEV_CON_ID("scr",		&mstp_clks[MSTP319]), /* SCR */
+	CLKDEV_DEV_ID("r8a7778-vin.0", &mstp_clks[MSTP110]), /* VIN0 */
+	CLKDEV_DEV_ID("r8a7778-vin.1", &mstp_clks[MSTP109]), /* VIN1 */
 	CLKDEV_DEV_ID("gp2d",		&mstp_clks[MSTP107]), /* GP2D */
 	CLKDEV_CON_ID("sgx",		&mstp_clks[MSTP105]), /* SGX */
 	CLKDEV_DEV_ID("rcar-du.0",	&mstp_clks[MSTP103]), /* DU */
diff --git a/arch/arm/mach-shmobile/include/mach/r8a7778.h b/arch/arm/mach-shmobile/include/mach/r8a7778.h
index 35dfdb6..93225d8 100644
--- a/arch/arm/mach-shmobile/include/mach/r8a7778.h
+++ b/arch/arm/mach-shmobile/include/mach/r8a7778.h
@@ -2,6 +2,10 @@
 #define __ASM_R8A7778_H__
 
 #include <linux/sh_clk.h>
+#include <linux/platform_data/camera-rcar.h>
+
+extern void r8a7778_add_vin_device(int id,
+				   struct rcar_vin_platform_data *pdata);
 
 /* Pin Function Controller:
  * GPIO_FN_xx - GPIO used to select pin function
diff --git a/arch/arm/mach-shmobile/setup-r8a7778.c b/arch/arm/mach-shmobile/setup-r8a7778.c
index 760f438..b8a3de3 100644
--- a/arch/arm/mach-shmobile/setup-r8a7778.c
+++ b/arch/arm/mach-shmobile/setup-r8a7778.c
@@ -36,6 +36,7 @@
 #include <linux/sh_timer.h>
 #include <linux/uio_driver.h>
 #include <linux/dma-contiguous.h>
+#include <linux/dma-mapping.h>
 #include <mach/hardware.h>
 #include <mach/irqs.h>
 #include <mach/r8a7778.h>
@@ -515,6 +516,39 @@ static struct platform_device *r8a7778_early_devices[] __initdata = {
 static struct platform_device *r8a7778_late_devices[] __initdata = {
 };
 
+/* VIN */
+#define R8A7778_VIN(idx) \
+static struct resource vin##idx##_resources[] = {			\
+	DEFINE_RES_MEM(0xffc50000 + 0x1000 * (idx), 0x1000),		\
+	DEFINE_RES_IRQ(gic_iid(0x5a)),					\
+};									\
+									\
+static struct platform_device_info vin##idx##_info __initdata = {	\
+	.parent		= &platform_bus,				\
+	.name		= "r8a7778-vin",				\
+	.id		= idx,						\
+	.res		= vin##idx##_resources,				\
+	.num_res	= ARRAY_SIZE(vin##idx##_resources),		\
+	.dma_mask	= DMA_BIT_MASK(32),				\
+}
+
+R8A7778_VIN(0);
+R8A7778_VIN(1);
+
+static struct platform_device_info *vin_info_table[] __initdata = {
+	&vin0_info,
+	&vin1_info,
+};
+
+void __init r8a7778_add_vin_device(int id, struct rcar_vin_platform_data *pdata)
+{
+	BUG_ON(id < 0 || id > 1);
+
+	vin_info_table[id]->data = pdata;
+	vin_info_table[id]->size_data = sizeof(*pdata);
+	platform_device_register_full(vin_info_table[id]);
+}
+
 void __init r8a7778_add_standard_devices(void)
 {
 #ifdef CONFIG_CACHE_L2X0
-- 
1.8.2

